AsyncDisplayKit/ASImageNode+AnimatedImage.mm

```diff
@@ -0,0 +1,226 @@
+
+#pragma mark - GIF support
+
+- (void)setAnimatedImage:(id <ASAnimatedImageProtocol>)animatedImage
+{
+  ASDN::MutexLocker l(_animatedImageLock);
+  if (!ASObjectIsEqual(_animatedImage, animatedImage)) {
+    _animatedImage = animatedImage;
+  }
+  if (animatedImage != nil) {
+    __weak ASImageNode *weakSelf = self;
+    if ([animatedImage respondsToSelector:@selector(setCoverImageReadyCallback:)]) {
+      animatedImage.coverImageReadyCallback = ^(UIImage *coverImage) {
+        [weakSelf coverImageCompleted:coverImage];
+      };
+    }
+    
+    animatedImage.playbackReadyCallback = ^{
+      [weakSelf animatedImageFileReady];
+    };
+  }
+}
+
+- (id <ASAnimatedImageProtocol>)animatedImage
+{
+  ASDN::MutexLocker l(_animatedImageLock);
+  return _animatedImage;
+}
+
+- (void)setAnimatedImagePaused:(BOOL)animatedImagePaused
+{
+  ASDN::MutexLocker l(_animatedImagePausedLock);
+  _animatedImagePaused = animatedImagePaused;
+  ASPerformBlockOnMainThread(^{
+    if (animatedImagePaused) {
+      [self stopAnimating];
+    } else {
+      [self startAnimating];
+    }
+  });
+}
+
+- (BOOL)animatedImagePaused
+{
+  ASDN::MutexLocker l(_animatedImagePausedLock);
+  return _animatedImagePaused;
+}
+
+- (void)coverImageCompleted:(UIImage *)coverImage
+{
+  BOOL setCoverImage = YES;
+  {
+    ASDN::MutexLocker l(_displayLinkLock);
+    if (_displayLink != nil && _displayLink.paused == NO) {
+      setCoverImage = NO;
+    }
+  }
+  
+  if (setCoverImage) {
+    self.image = coverImage;
+  }
+}
+
+- (void)animatedImageFileReady
+{
+  dispatch_async(dispatch_get_main_queue(), ^{
+    [self startAnimating];
+  });
+}
+
+- (void)startAnimating
+{
+  ASDisplayNodeAssertMainThread();
+  if (ASInterfaceStateIncludesVisible(self.interfaceState) == NO) {
+    return;
+  }
+  
+  if (self.animatedImagePaused) {
+    return;
+  }
+  
+  if (self.animatedImage.playbackReady == NO) {
+    return;
+  }
+  
+#if ASAnimatedImageDebug
+  NSLog(@"starting animation: %p", self);
+#endif
+  ASDN::MutexLocker l(_displayLinkLock);
+  if (_displayLink == nil) {
+    _playHead = 0;
+    _displayLink = [CADisplayLink displayLinkWithTarget:[ASWeakProxy weakProxyWithTarget:self] selector:@selector(displayLinkFired:)];
+    _displayLink.frameInterval = self.animatedImage.frameInterval;
+    
+    [_displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
+  } else {
+    _displayLink.paused = NO;
+  }
+}
+
+- (void)stopAnimating
+{
+#if ASAnimatedImageDebug
+  NSLog(@"stopping animation: %p", self);
+#endif
+  ASDisplayNodeAssertMainThread();
+  ASDN::MutexLocker l(_displayLinkLock);
+  _displayLink.paused = YES;
+  self.lastDisplayLinkFire = 0;
+  
+  [self.animatedImage clearAnimatedImageCache];
+}
+
+- (void)visibilityDidChange:(BOOL)isVisible
+{
+  [super visibilityDidChange:isVisible];
+  
+  ASDisplayNodeAssertMainThread();
+  if (isVisible) {
+    if (self.animatedImage.coverImageReady) {
+      self.image = self.animatedImage.coverImage;
```

Comment:
Setting self.image still results in an async display pass, even if the image is already decoded...because of the way _ASDisplayLayer integrates with CA display.  We may want to force a synchronous display here.  It's tricky though because it is possible in certain timings that layout may not be fully applied yet here (e.g. -setNeedsLayout called, but -layoutSubviews might run next, depending on whether the range controller dispatches to the next runloop - it does dispatch today, but in the future we don't want it to).

All around complex shittiness here, let's discuss more soon.  This isn't a big problem for now.